src/crimson/common/config_proxy.h

```diff
@@ -34,25 +35,33 @@ class ConfigProxy : public seastar::peering_sharded_service<ConfigProxy>
   }
 
   // apply changes to all shards
-  // @param func a functor which accepts @c "ConfigValues&", and returns
-  //             a boolean indicating if the values is changed or not
+  // @param func a functor which accepts @c "ConfigValues&"
   template<typename Func>
   seastar::future<> do_change(Func&& func) {
-    auto new_values = seastar::make_lw_shared(*values);
-    func(*new_values);
-    return seastar::do_with(seastar::make_foreign(new_values),
-			    [this](auto& foreign_values) {
-      return container().invoke_on_all([&foreign_values](ConfigProxy& proxy) {
-        return foreign_values.copy().then([&proxy](auto foreign_values) {
-	  proxy.values.reset();
-	  proxy.values = std::move(foreign_values);
-	  proxy.obs_mgr.apply_changes(proxy.values->changed,
-				      proxy, nullptr);
+    return container().invoke_on(values.get_owner_shard(),
+                                 [func = std::move(func)](ConfigProxy& owner) {
+      // always apply the new settings synchronously on the owner shard, to
+      // avoid racings with other do_change() calls in parallel.
+      owner.values.reset(seastar::make_lw_shared(*owner.values));
+      func(*owner.values);
+      // instead of passing `values` by reference to we need to pass a copy of
+      // it, otherwise the `values` instance could be overwritten by succeeding
+      // calls of do_change(), and other shards will be copying a non-existing
+      // `values`.
+      return owner.values.copy().then([&owner](auto copy_values) {
+        return owner.container().invoke_on_others(
+          [&values = copy_values](ConfigProxy& proxy) {
+            return values.copy().then([&proxy](auto foreign_values) {
+              proxy.values.reset();
+              proxy.values = std::move(foreign_values);
```

Comment:
`foreign_ptr` makes things a bit awkward here. first, capturing `[&values = copy_values]` by reference doesn't look safe because `copy_values` goes out of scope. second, calling `values.copy()` from other shards requires a round-trip to shard0 and back to increment the refcount

we can work around both of these issues by using the `lw_shared_ptr` for local continuations, then creating a separate `foreign_ptr` for each shard. for example:
```c++
  template<typename Func>
  seastar::future<> do_change(Func&& func) {
    return container().invoke_on(0u, [func = std::move(func)](auto& owner) {
      // apply the changes to a copy of the values
      auto new_values = seastar::make_lw_shared(*owner.values);
      func(*new_values);

      // always apply the new settings synchronously on the owner shard, to
      // avoid racings with other do_change() calls in parallel.
      owner.values.reset(new_values);

      return seastar::parallel_for_each(boost::irange(1u, seastar::smp::count),
                                        [&owner, new_values] (auto cpu) {
        return owner.container().invoke_on(cpu,
          [foreign_values = seastar::make_foreign(new_values)](ConfigProxy& proxy) mutable {
            proxy.values.reset();
            proxy.values = std::move(foreign_values);
            proxy.obs_mgr.apply_changes(proxy.values->changed,
                                        proxy, nullptr);
          });
        }).then([&owner] {
          owner.values->changed.clear();
        });
      });
  }
```
what do you think?

Reply:
and i guess that last part should use new_values instead?
```c++
        }).then([new_values] {
          new_values->changed.clear();
        });
```

Reply:
i think it also needs to clear `new_values->changed` before calling `func()` so the observers only see the values changed by this operation?
```c++
      // apply the changes to a copy of the values
      auto new_values = seastar::make_lw_shared(*owner.values);
      new_values->changed.clear();
      func(*new_values);
```

Reply:
yeah, the leftover changes should be taken care of the continuation of the operation that caused it.

Reply:
thanks @cbodley ! i was keen on using the `invoke_on_other()` method, as it looked like a perfect fit for this case. i even considered using `map_reduce() but yeah, we need to construct a different `foreign_ptr<>` for each shards on the owner shard. the most straightforward is to use `parallel_for_each()` + `invoke_on()`, i.e. the combination of "local + remote", none of the `sharded<>` methods offers us this flexibility.